### DISCUSSION TITLE
[Feature] JustifiedLayout

### DISCUSSION THREAD

**Discussion by Poker-sang** [CommunityToolkit/Labs-Windows#435] [Experiments]
A virtualizing layout that justifies content.

It wraps like `WrapLayout`, but with some new features:

* It makes all controls at the same height (via the RowHeight property)
* It moderately changes the aspect ratio of the controls so that each row of controls is the same length. But it does not change the aspect ratio too much.

Windows Gallery:

![image](https://github.com/CommunityToolkit/Labs-Windows/assets/62325494/50fa7f48-192f-4037-bb5d-f5758d2b0da5)

I have implemented a simple one: [JustifiedLayout](https://github.com/Poker-sang/JustifiedLayout)

![image](https://github.com/CommunityToolkit/Labs-Windows/assets/62325494/db429c53-25c5-4db2-b032-fe57d007462a)

**Comment by niels9001** [comment]
@Poker-sang Thanks for opening up this discussion!

@michael-hawker @Arlodotexe Thoughts?

**Comment by kmgallahan** [comment]
@niels9001 @Poker-sang Already being made available in muxc as `RiverFlowLayout`:

https://youtu.be/Fj4YjgDqm84?t=504

Only available in experimental releases at this time.

**Reply by Poker-sang** [reply]
Thanks! 😃 I was just confused, for my implemention has dozens of bugs. I can't wait to try RiverFlowLayout. Is it available in the recent pre-release nuget package?

**Reply by kmgallahan** [reply]
https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/experimental-channel#version-13-experimental-130-experimental1

https://www.nuget.org/packages/Microsoft.WindowsAppSDK/

**Reply by Poker-sang** [reply]
Thanks a lot!

**Reply by Poker-sang** [reply]
@kmgallahan Sorry to bother you again. Does the WASDK provide a layout similar to AdaptiveGridView? If you just put controls of the same width in RiverFlowLayout, it can sometimes make the columns unaligned. Should I start a new discussion seeking AdaptiveGridLayout?

**Reply by kmgallahan** [reply]
I'm unsure of the exact behavior you are looking for and what combination of settings and layouts you've already tried.

Have you already gone through the WinUI 3 Gallery & Windows Community Toolkit Sample App?

Generally, the community toolkit doesn't include things that duplicate functionality from muxc.

**Reply by Poker-sang** [reply]
Yes I have checked the Gallery & Toolkit App and there is nothing I want.😢 I would like a layout that automatically adjusts the width of each item, yet ensures that the items are aligned to the grid. It's like AdaptiveGridView, but I need a version of the layout. (RiverFlowLayout doesn't do this very well)

**Reply by Poker-sang** [reply]
BTW this is not the JustifiedLayout I was talking about. that problem has been solved very well by RiverFlowLayout, thanks for your answer!☺️

**Comment by Poker-sang** [comment]
Looks like only collaborators have access to mark this as an answer.

**Comment by Poker-sang** [comment]
@kmgallahan It performs like this:
![GIF](https://github.com/CommunityToolkit/Labs-Windows/assets/62325494/5c732589-c3dc-4401-92a2-22338f103d95)
However, this is the performance of RiverFlowLayout:
![GIF2](https://github.com/CommunityToolkit/Labs-Windows/assets/62325494/5a2e2225-fc95-4a6e-b49a-4fbaba9b332e)

**Reply by kmgallahan** [reply]
So you just want all widths and heights to be the same, but don't want to use AdaptiveGridView?

**Reply by Poker-sang** [reply]
@kmgallahan Yes! I would like to use layout, for ItemsRepeater can switch among the layouts easily.

**Reply by kmgallahan** [reply]
What about [`UniformGridLayout`](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.uniformgridlayout) with `ItemsStretch` = `Uniform` and `MinItemHeight` + `MinItemWidth` set, and the (presumably) `Image`s set to use one of the `Stretch` modes and a max width & height.

**Reply by Poker-sang** [reply]
@kmgallahan oh that's it! That's exactly what I want. Thank you!!! (p.s. `ItemsStretch` should be set `Fill`)
